New SMI extension, which includes the `smi-adaptor`

Pothulapati opened the pull request:
This PR adds the new SMI Extension, which includes the
`CLI` and the `smi-adaptor`. Along with these components, 
scripts that are needed to build, run and test them are also
added. Patterns from the linkerd2 repo are followed, without
inheriting unnecessary complexity.

## `smi-adaptor` controller

Runs in a loop and listens for `TrafficSplit` Resources
and makes sure that the relevant `ServiceProfile` state is present.
This follows the same pattern as that of the [`k8s/sample-controller`](https://github.com/kubernetes/sample-controller)

The state changes are as follows:

- User creates or updates a `TS` resource
    - No relevant `SP` already present
       - **Create New `SP`**
    - Relevant `SP` already present
      - `SP` contains the `smi.linkerd.io/skip` annotation
        - **Skip**
      - No `smi.linkerd.io/skip` Annotation
        - **Override `SP.dstOverrides` field**
- User deletes a `TS` resource
   - **empty `dstOverrides` field**

## Building and Running

First, Install a Linkerd enabled Kubernetes cluster.
Then, Run the following commands:
```bash
# Build the CLi
./bin/build-cli-bin
# Build the smi-adaptor docker image
./bin/docker-build
# Load the image into kind
kind load docker-image cr.l5d.io/linkerd/smi-adaptor:$(./bin/root-tag)
# Run the smi install cmd
./target/cli/linux-amd64/linkerd-smi install | k apply -f -
```

alpeb commented on bin/docker-build-smi-adaptor:
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
```
I think you can just rename this file `bin/docker-build` and avoid the indirection

Pothulapati replied:
My reason for doing this, was as we might have a docker build file for CLI in the near future i.e once we add release logic 🤔

Pothulapati replied:
Simplified this!

alpeb commented on bin/kind:
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env sh
```
Most of us being on k3d now, you could get rid of this one...

Pothulapati replied:
I use kind a lot 😬 . I usually get timeouts a lot in k3d for some reason 🤔

alpeb replied:
Ah must be windows :-P
Good then! :-)

alpeb commented on go.mod:
```diff
@@ -0,0 +1,14 @@
+
+require (
+	github.com/elazarl/goproxy/ext v0.0.0-20210110162100-a92cc753f88e // indirect
+	github.com/linkerd/linkerd2 v0.5.1-0.20210518142200-847b7441831a
```
why `v0.5.1`?

Pothulapati replied:
That's the latest version. Currently the linkerd pkgs are not versioned and published to go pkgs i.e https://pkg.go.dev/github.com/linkerd/linkerd2. I think, the reason is because we don't use semver.

We instead directly depend on a specific commit (here one of the latest recent commit) , which go mod for some reason weirdly parsed as `v0.5.1` 🤔

adleong replied:
Since we don't publish go module semver, you should depend on pseudoversions instead.  The latest pseudoversion is:

```console
> pseudoversion
v0.0.0-20210526154658-5cecf863e287
```

alpeb commented on charts/linkerd-smi/templates/NOTES.txt:
```diff
@@ -0,0 +1,8 @@
+
+To make sure everything works as expected, run the following:
+
+  linkerd smi check
```
I didn't see any `check` subcommand here... :mag:

alpeb commented on charts/linkerd-smi/values.yaml:
```diff
@@ -0,0 +1,11 @@
+namespace: smi
```
Why not `linkerd-smi`, following what we've done with the other extensions?

alpeb commented on bin/k3d:
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env sh
```
This file requires the executable flag

alpeb reviewed (commented):
I mainly checked the scaffolding and left a few minor notes. I'll give then a go at the adaptor and let you know :wink:

olix0r commented:
It's not a hard blocker, but it would be great if we didn't have to promulgate all of our crufty bin scripts. Things like the docker build scripts don't seem particularly necessary, especially if this repo has a single dockerfile. Rather than import all of the scripts and actions whole-hog, can we start with only the things we absolutely need and then build up from there.  For instance, why do we want bin/kind to be checked in at all. If you have kind installed, you can just use kind without using our wrapper scripts, surely?

olix0r commented:
It's not a hard blocker, but it would be great if we didn't have to promulgate all of our crufty bin scripts. Things like the docker build scripts don't seem particularly necessary, especially if this repo has a single dockerfile. Rather than import all of the scripts and actions whole-hog, can we start with only the things we absolutely need and then build up from there.  For instance, why do we want bin/kind to be checked in at all. If you have kind installed, you can just use kind without using our wrapper scripts, surely?

Pothulapati commented:
@olix0r The docker side of things is a really good suggestion but for others i.e `kind`, `helm`, etc scripts that are added, They would be really useful in the CI. We are already using some of these i.e `markdownlint`, etc in the PR's workflow! I've tried to be very minimal here, and moved only the ones that we absolutely need rn except for the `kind`, `k3d`, `helm` which are useful with the integration test's CI!

The reason for adding them is that, it would establish a pattern that we can use it test it currently, but also use the same in the testing and release workflows! 🤔

adleong commented:
WDYT about splitting this into two PRs?  One which sets up the repo (sets up READMEs, licences, build scripts, CI, etc. etc.) and a separate PR which adds the smi adaptor?

adleong commented on adaptor/main.go:
```diff
@@ -0,0 +1,98 @@
+
+	log.Info("Using cluster domain: ", *clusterDomain)
+
+	if *traceCollector != "" {
```
Does this component need tracing?  All it does is interact with the k8s API.  Doesn't seem like tracing is useful.

adleong commented on bin/image-load:
```diff
@@ -0,0 +1,102 @@
+do
+  case $1 in
+    -h|--help)
+      echo "Load into KinD/k3d the images for Linkerd's proxy, controller, metrics-api, web, grafana, cli-bin, tap, debug and cni-plugin."
```
This seems unnecessary

Pothulapati commented:
@adleong  Makes sense. Raised #2 to tackle the boilerplate, and will update this PR to be specifically about SMI adaptor. WDYT? 🤔

adleong commented on pkg/adaptor/controller.go:
```diff
@@ -0,0 +1,365 @@
+	return ok
+}
+
+func equal(spA *serviceprofile.ServiceProfile, spB *serviceprofile.ServiceProfile) bool {
```
I'll echo what I said on the other PR: I don't think that there's a significant cost to doing redundant updates so I'd be inclined to skip the equals check.

Pothulapati commented:
@adleong Reg Pseudoversions, Yes We currently do that. That `v0.5.1` prefix is just the last released tag that go mod automatically prefixes https://golang.org/ref/mod#pseudo-versions

I've updated it to `v0.0.0` to prevent confusion!

Pothulapati commented:
Okay, Updated this PR to have working integration tests. The current integration tests were moved over from https://github.com/linkerd/linkerd2/pull/6097 where we were checking if the stats cmd displayed the right metrics after installing a trafficsplit.

But as the SMI Extension is specifically about converting `TrafficSPlit` into `ServiceProfiles`, I will be updating the integration tests to check for the same i.e the state of serviceprofiles for a relevant TrafficSplit

Feel free to review the controller logic, etc meanwhile!

alpeb commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,39 @@
+      # Unpinned action version so that we automatically get analyzer updates.
+      uses: github/codeql-action/init@v1
+      with:
+        languages: ${{ matrix.language }}
```
You can simplify by using `go` here and removing the `matrix` entry above

alpeb commented on adaptor/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+WORKDIR /linkerd-smi-build
+COPY go.mod go.sum ./
+RUN go mod download
+ARG TARGETARCH
```
`TARGETARCH` is not used. In linkerd2's repo we use it in `bin/install-deps` but given the size of this repo, this might not be needed. I'd say let's start without using this `go-deps` stage at all, and then think about adding it only if the whole build is taking a lot of time.

Pothulapati replied:
Makes sense. Updated it to be a simple 2 stage build i.e building, and Copying the output binary!

alpeb commented on adaptor/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+
+## package runtime
+FROM scratch
+ARG LINKERD_VERSION
```
Where is `LINKERD_VERSION` used?

alpeb commented on charts/linkerd-smi/values.yaml:
```diff
@@ -0,0 +1,11 @@
+  image: 
+    name: cr.l5d.io/linkerd/smi-adaptor
+    version: *linkerd_smi_version
+    pullPolicy: IfNotPresent
```
I didn't see `pullPolicy` being used

alpeb commented on test/install_test.go:
```diff
@@ -0,0 +1,303 @@
+	// Install Linkerd
+	out, err := TestHelper.LinkerdRun("install")
+	if err != nil {
+		linkerdtestutil.AnnotatedFatal(t, "'linkerd smi install' command failed", err)
```
```suggestion
		linkerdtestutil.AnnotatedFatal(t, "'linkerd install' command failed", err)
```

alpeb commented on test/install_test.go:
```diff
@@ -0,0 +1,303 @@
+	// Install Viz extension
+	out, err = TestHelper.LinkerdRun("viz", "install")
+	if err != nil {
+		linkerdtestutil.AnnotatedFatal(t, "'linkerd smi install' command failed", err)
```
```suggestion
		linkerdtestutil.AnnotatedFatal(t, "'linkerd viz install' command failed", err)
```

alpeb replied:
Marking back as unresolved :wink:

Pothulapati replied:
Yep, As these tests are being re-worked, I marked all of them as resolved!

alpeb commented on .github/workflows/integration_tests.yml:
```diff
@@ -0,0 +1,104 @@
+    - '*.md'
+    - '**/*.md'
+    branches:
+    - main
```
Would you add here the minimal permissions?
```
permissions:
    contents: read
````

alpeb commented on .github/workflows/integration_tests.yml:
```diff
@@ -0,0 +1,104 @@
+      with:
+        name: build-archives
+    - name: Create k8s Kind Cluster
+      uses: helm/kind-action@v1.1.0
```
Any chance to get something similar but for k3d? Startup time is much faster if I'm not mistaking.

alpeb commented on test/install_test.go:
```diff
@@ -0,0 +1,303 @@
+	}
+
+	// Install SMI extension
+	out, err = TestHelper.LinkerdSMIRun("install")
```
I think it'd be good to wait for viz' deployments rollouts before installing smi, so things are more deterministic across CI runs

alpeb commented on test/install_test.go:
```diff
@@ -0,0 +1,303 @@
+	}
+
+	version := "v1alpha1-with-smi-adaptor"
+	prefixedNs := version
```
Is there need for this extra var?

alpeb reviewed (commented):
Thanks for the updates @Pothulapati :+1: 

I still haven't yet looked into the controller code itself, but here are a few comments about everything else:

- Slipped through the previous merged PR: Better not ignore md files in the static_checks, given we're gonna try linting those files :wink: 
- Still to do the release CI workflow?
- Can you add a subcommand `version`? I guess `check` will come later, but `version` should be doable here...
- I'm getting a few errors when running `bin/build-cli-bin` indicating `go.sum` needs to be updated:
```
$ bin/build-cli-bin 
generate.go:7:2: missing go.sum entry for module providing package github.com/shurcooL/vfsgen; to add:
        go mod download github.com/shurcooL/vfsgen
pkg/static/templates.go:1: running "go": exit status 1

$ bin/build-cli-bin 
/home/alpeb/go/pkg/mod/github.com/shurcoo!l/vfsgen@v0.0.0-20181202132449-6a9ea43bcacd/generator.go:18:2: missing go.sum entry for module providing package github.com/shurcooL/httpfs/vfsutil; to add:
        go mod download github.com/shurcooL/httpfs
pkg/static/templates.go:1: running "go": exit status 1
```

> But as the SMI Extension is specifically about converting TrafficSPlit into ServiceProfiles, I will be updating the integration tests to check for the same i.e the state of serviceprofiles for a relevant TrafficSplit

Looking forward to this :+1:

Pothulapati commented:
@alpeb 

> Slipped through the previous merged PR: Better not ignore md files in the static_checks, given we're gonna try linting those files :wink: 

Good Find. Updated!

> Still to do the release CI workflow?

Not Yet, Will be doing this as a separate PR

> Can you add a subcommand `version`? I guess `check` will come later, but `version` should be doable here...

Yep, Makes sense. Updated.
> I'm getting a few errors when running `bin/build-cli-bin` indicating `go.sum` needs to be updated:

Fixed

alpeb commented on adaptor/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+RUN go mod download
+COPY adaptor adaptor
+COPY pkg pkg
+RUN CGO_ENABLED=0 GOOS=linux go build -o /out/smi-adaptor -tags production -mod=readonly -ldflags "-s -w" ./adaptor
```
I commented earlier about `TARGETARCH` not being used int the former `go-deps` stage. But it's still required in this golang stage in order to support multiarch builds through the `GOARCH` env var.

Pothulapati replied:
Ahh, Right. I was in the impression that we can add it later but seems too simple to add it now!

Pothulapati commented:
@alpeb Updated the integration tests. Now, we check for the conversion i.e SP should have specific values based on the updates to TS.

adleong commented on adaptor/main.go:
```diff
@@ -0,0 +1,91 @@
+		log.Fatalf("Error building example clientset: %s", err.Error())
+	}
+
+	spInformerFactory := spinformers.NewSharedInformerFactory(spClient, time.Second*30)
```
is this unused?

Pothulapati replied:
Good Find. Removed!

adleong commented on bin/_docker.sh:
```diff
@@ -0,0 +1,101 @@
+    echo "$name"
+}
+
+docker_build() {
```
this seems like a lot of scripting; is it all necessary?  can we just inline the docker command where we need it?

adleong commented on bin/helm-build:
```diff
@@ -0,0 +1,52 @@
+        echo 'Version tag is malformed'
+        exit 1
+    fi
+    fullVersion=${BASH_REMATCH[0]}
```
Is this complexity coming from the fact that we're using the same versioning scheme as Linkerd?  could we simplify this if this project just used a simple semver versioning?

It's likely that the smi-adaptor will release on a different schedule than Linkerd (probably much less frequently) and compatibility between the two projects will only be affected when the ServiceProfile CRD changes.  So I don't think we want to follow the edge/stable versioning scheme here.

adleong commented on charts/linkerd-smi/README.md:
```diff
@@ -0,0 +1,76 @@
+# linkerd-smi
+
+The Linkerd-SMI extension adds SMI adaptor to the Linkerd install
```
We can follow up in a later PR, but we probably want to flesh this out to at least say what the SMI adaptor is or does.

Pothulapati replied:
Yep, Currently the Helm Readme is unused. A separate PR will be put to add Helm support for the SMI Extension. I will address the readme update in that with more explanations!

adleong commented on charts/linkerd-smi/templates/rbac.yaml:
```diff
@@ -0,0 +1,43 @@
+rules:
+- apiGroups: ["linkerd.io"]
+  resources: ["serviceprofiles"]
+  verbs: ["list", "get", "watch", "create", "update"]
```
super minor but technically we don't need "watch" on SP, right?

Pothulapati replied:
Yes, Removed

adleong commented on charts/linkerd-smi/templates/rbac.yaml:
```diff
@@ -0,0 +1,43 @@
+  namespace: {{.Values.namespace}}
+  labels:
+    linkerd.io/extension: smi
+    component: smi-adaptor
```
newline

adleong commented on charts/linkerd-smi/values.yaml:
```diff
@@ -0,0 +1,11 @@
+
+installNamespace: true
+clusterDomain: cluster.local
+linkerdSMIVersion: &linkerd_smi_version linkerdSMIVersionValue
```
do we need this version template variable at all?  can we just treat the image version as a normal helm value?

adleong replied:
I think we discussed that the linkerdSMIVersionValue placeholder can go directly in the version property without creating this redirect.

adleong commented on cli/cmd/testdata/install_default.golden:
```diff
@@ -0,0 +1,86 @@
+  namespace: linkerd-smi
+  labels:
+    linkerd.io/extension: smi
+    component: smi-adaptor
```
newline

adleong commented on cli/cmd/testdata/install_override_namespace.golden:
```diff
@@ -0,0 +1,86 @@
+  namespace: linkerd-smi-2
+  labels:
+    linkerd.io/extension: smi
+    component: smi-adaptor
```
newline

adleong commented on pkg/adaptor/controller.go:
```diff
@@ -0,0 +1,365 @@
+		// Check if there is a relevant SP that was created or updated by SMI Controller
+		// and clean up its dstOverrides
+		log.Infof("trafficsplit/%s is deleted, trying to cleanup the relevant serviceprofile", name)
+		sp, err := c.spclientset.LinkerdV1alpha2().ServiceProfiles(namespace).Get(ctx, c.toFQDN(service, namespace), metav1.GetOptions{})
```
If the err here is NotFound then I think we can just return nil since the state is what it should be.

Pothulapati replied:
Makes sense!

adleong commented on pkg/adaptor/controller.go:
```diff
@@ -0,0 +1,365 @@
+// enqueueTS takes a Ts resource and converts it into a key
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than TS.
+func (c *SMIController) enqueueTS(obj interface{}) {
```
let's enforce type safety as early possible.  can we have this function take a TrafficSplit as an argument and have the informer handler func do casting and error handling?

Pothulapati replied:
Makes sense. Updated!

adleong commented on pkg/adaptor/controller.go:
```diff
@@ -0,0 +1,365 @@
+
+// trafficSplitKeyFunc takes a TS, and gives back a
+// namespace/name/service key string
+func trafficSplitKeyFunc(obj interface{}) (string, error) {
```
Instead of using a string for the key, we should use a (namespace, name, service) struct.  
Or, even better, can we just use the TS itself as the key?

Pothulapati replied:
We can do that but I don't see the benefit here, The purpose of the key being just a string(we need namespace, name and service to match with the relevant serviceprofile from the trafficsplit) is that, we can move the event around without having to worry about the object. and retrieve the actual TS when we need it i.e `syncHandler` so that we also get the then state of the object? 🤔

adleong reviewed (commented):
As a general comment, there's still a lot of build related infrastructure here.  It definitely feels like we copied over everything from Linkerd and then tried to remove what we could.  I think a better approach would be to do the opposite: start with nothing and add the most minimal build scripting that we can, only pulling stuff from Linkerd where necessary to avoid reinventing the wheel.  This is a simple project and I suspect we can get away with far less tooling.

The code itself looks pretty good!  I will do some more on-cluster testing of this today.

dadjeibaah commented on pkg/adaptor/controller.go:
```diff
@@ -0,0 +1,365 @@
+	workqueue workqueue.RateLimitingInterface
+}
+
+// NewController returns a new sample controller
```
```suggestion
// NewController returns a new SMI controller
```

adleong commented on pkg/adaptor/controller.go:
```diff
@@ -0,0 +1,365 @@
+
+// toServiceProfile converts the given TrafficSplit into the relevant ServiceProfile resource
+func (c *SMIController) toServiceProfile(ts *trafficsplit.TrafficSplit) *serviceprofile.ServiceProfile {
+	spResource := serviceprofile.ServiceProfile{
```
I think you need to set Routes to an empty list (instead of nil) in order to avoid validation errors:

```
time="2021-06-09T20:45:25Z" level=error msg="error syncing 'booksapp/error-split/books': ServiceProfile.linkerd.io \"books.booksapp.svc.cluster.local\" is invalid: spec.routes: Invalid value: \"null\": spec.routes in body must be of type array: \"null\", requeuing"
```

Pothulapati replied:
Yep, That is the case on stable, but that restriction is removed on the edge. Are you testing on the edge? 🤔 

I've tried to not touch other fields except for `dstOverrides` to keep the logic simple. Based on our discussions about supporting < `2.10.2`, I can work on fixing this.

dadjeibaah commented on pkg/adaptor/controller.go:
```diff
@@ -0,0 +1,365 @@
+	// Get the Ts resource with this namespace/name
+	ts, err := c.tsclientset.SplitV1alpha1().TrafficSplits(namespace).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
```
Can we include a similar comment here like how we have it on line 217?

adleong reviewed (commented):
In the steady state where no resources are changing, the smi-adaptor seems to fire and resync every 30 seconds:

```
time="2021-06-09T21:15:52Z" level=info msg="serviceprofile/books.booksapp.svc.cluster.local already present"
time="2021-06-09T21:15:52Z" level=info msg="Successfully synced 'booksapp/error-split/books'"
time="2021-06-09T21:16:22Z" level=info msg="serviceprofile/books.booksapp.svc.cluster.local already present"
time="2021-06-09T21:16:22Z" level=info msg="Successfully synced 'booksapp/error-split/books'"
time="2021-06-09T21:16:52Z" level=info msg="serviceprofile/books.booksapp.svc.cluster.local already present"
time="2021-06-09T21:16:52Z" level=info msg="Successfully synced 'booksapp/error-split/books'"
time="2021-06-09T21:17:22Z" level=info msg="serviceprofile/books.booksapp.svc.cluster.local already present"
time="2021-06-09T21:17:22Z" level=info msg="Successfully synced 'booksapp/error-split/books'"
```

is this expected?

dadjeibaah commented on pkg/adaptor/controller.go:
```diff
@@ -0,0 +1,365 @@
+// updateDstOverrides updates the dstOverrides of the given serviceprofile
+// to match that of the trafficsplit
+func updateDstOverrides(sp *serviceprofile.ServiceProfile, ts *trafficsplit.TrafficSplit, clusterDomain string) {
+	sp.Spec.DstOverrides = []*serviceprofile.WeightedDst{}
```
TIOLI: Since we already know that `len(sp.Spec.DstOverrides)` is going to be equal to the `len(ts.Spec.Backends)`, can we just initialize `sp.Spec.DstOverrides` with the same size array and just use the index to update the array in place?
```go
func updateDstOverrides(sp *serviceprofile.ServiceProfile, ts *trafficsplit.TrafficSplit, clusterDomain string) {
	sp.Spec.DstOverrides = make([]*serviceprofile.WeightedDst, len(ts.Spec.Backends))
	for i, backend := range ts.Spec.Backends {
		sp.Spec.DstOverrides[i] = &serviceprofile.WeightedDst{
			Authority: fqdn(backend.Service, ts.Namespace, clusterDomain),
			Weight:    *backend.Weight,
		}
	}
}
```

Pothulapati replied:
Makes Sense. Updated!

dadjeibaah reviewed (commented):
This looks great! I had a couple of comments and also it seems that the logs get a little cluttered about a minute after running the adaptor at info even with a single trafficsplit. Might I suggest moving some log lines to debug to reduce the amount of logs we get when in steady state.

Pothulapati commented:
Now, The controller is updated to work with all k8s versions and the recent Linkerd edges. This wasn't the case because of the `sp.Spec.Routes` field issue that @adleong mentioned, causing the adaptor to throw errors in k8s <= 1.19 versions. A commit to fix this is pushed.

The adaptor still won't work with Linkerd stable because of `spValidator` actively rejecting ServiceProfiles with `len(sp.Spec.Routes) == 0` at https://github.com/linkerd/linkerd2/blob/release/stable-2.10/pkg/profiles/profiles.go#L65 and throwing 
```bash
time="2021-06-15T11:06:20Z" level=error msg="error syncing 'linkerd-smi-app/backend-traffic-split/backend-svc': admission webhook \"linkerd-sp-validator.linkerd.io\" denied the request: ServiceProfile \"backend-svc.linkerd-smi-app.svc.cluster.local\" has no routes, requeuing"
```

This condition was removed in main, and hence things should work fine on the edges!

Pothulapati commented:
Update about the `sp.Spec.Routes` issue,

Raised https://github.com/linkerd/linkerd2/pull/6271 to address it in the linkerd2 repo, so that we don't have to do this unnecessary work of making sure `sp.Spec.Routes` is not nil. Will update this PR, and remove the last commit once that PR is merged!

alpeb commented on charts/linkerd-smi/templates/adaptor.yaml:
```diff
@@ -0,0 +1,34 @@
+        imagePullPolicy: {{.Values.adaptor.image.pullPolicy}}
+        name: smi-adaptor
+        ports:
+        - containerPort: 9999
```
According to `main.go` metrics are served on 9995

alpeb commented on pkg/adaptor/controller.go:
```diff
@@ -0,0 +1,364 @@
+func (c *SMIController) Run(stopCh <-chan struct{}) error {
+	defer c.workqueue.ShutDown()
+
+	// Start the informer factories to begin populating the informer caches
```
```suggestion
```

dadjeibaah commented on .github/workflows/integration_tests.yml:
```diff
@@ -0,0 +1,106 @@
+    strategy:
+      matrix:
+        # TODO: Test on stable version too
+        version: [ edge]
```
super tiny nit: maybe remove the extra space in the array.

Pothulapati replied:
This is not needed as its very hard to add support in stable, because the SP validator actively rejects SPs with empty `len(sp.Spec.Routes)`. So, Removing this strategy stuff altogether.

dadjeibaah commented on adaptor/main.go:
```diff
@@ -0,0 +1,88 @@
+	// Create SP and TS clientsets
+	spClient, err := spclientset.NewForConfig(config)
+	if err != nil {
+		log.Fatalf("Error building example clientset: %s", err.Error())
```
```suggestion
		log.Fatalf("Error building serviceprofile clientset: %s", err.Error())
```

dadjeibaah commented on adaptor/main.go:
```diff
@@ -0,0 +1,88 @@
+
+	tsClient, err := tsclientset.NewForConfig(config)
+	if err != nil {
+		log.Fatalf("Error building example clientset: %s", err.Error())
```
```suggestion
		log.Fatalf("Error building trafficsplit clientset: %s", err.Error())
```

dadjeibaah commented on cli/cmd/root.go:
```diff
@@ -0,0 +1,65 @@
+	smiCmd.AddCommand(newCmdInstall())
+	smiCmd.AddCommand(newCmdUninstall())
+	smiCmd.AddCommand(newCmdVersion())
+
```
Anyway we include autocompletion for `kubecontext` and `linkerd-namespace`? It looks like we pull in all the package libraries from Linkerd so we should be able to do that possibly.

Pothulapati replied:
Done!

dadjeibaah reviewed (commented):
Sorry for getting this in late, but here are a couple more suggestions for your consideration.

Pothulapati commented:
@adleong I was replicate the resync problem that you specified. I updated the controller to have 10 minutes as the resync period instead in the latest commit with the following reasoning!

```text
Whenever SharedInformerFactories are specified a resync period
has to be specified. A resync events plays back all the events held
in the infromer cache. Since, The controller uses a workqueue
for processing and requeues any events that caused errors, This
resync period is largely not useful and is necessary to compensate
any bugs in the controller that are missed by the requeuing logic
in the workqueue.

This means that higher period is always advised, as its rare
to miss events. Hence, this period is being set to 10 minutes.

10 minutes is also the same period that the destination and other
controllers use in the core linkerd control-plane.

More on this: https://groups.google.com/g/kubernetes-sig-api-machinery/c/PbSCXdLDno0/m/v9gH3HXVDAAJ
```

dadjeibaah reviewed (approved):
Changes look good to me modulo other reviewers comments. I assume the machinery around building the CLI binary will be a follow up to this PR?

dadjeibaah commented:
Talked to @Pothulapati, In an offline convo and we already have scripts to build the CLI already. It looks like I just missed that in the `/bin` directory script. It would be good, however, to include the building of the CLI binary when we run `bin/docker-build`. This can probably be done in a follow-up PR. Maybe when we set up the repo to release binary images.

Pothulapati commented:
@adleong Just updated the PR to also remove the unnecessary equals check during SP updates like you suggested!

(This was already part of a different commit, which got reverted :/)

adleong reviewed (approved):
Thanks for all your hard work on this, @Pothulapati! Looks sweet!

alpeb reviewed (approved):
Looks good to me! :+1:
